# Project 3.2 - Train, Test, Split & Modelling

# Train Test & Split time
(oh and feature engineering on X_train)

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

##Train test splitting from sklearn
from sklearn.model_selection import train_test_split

##Stats model for linreg
import statsmodels.api as sm
import statsmodels.tools

from sklearn.preprocessing import StandardScaler, RobustScaler, MinMaxScaler

In [3]:
df=pd.read_csv("Life Expectancy Data Updated.csv")

In [13]:
df.head()

Country          Region  Year  Infant_deaths  Under_five_deaths  \
0  Turkiye     Middle East  2015           11.1               13.0   
1    Spain  European Union  2015            2.7                3.3   
2    India            Asia  2007           51.5               67.9   
3   Guyana   South America  2006           32.8               40.5   
4   Israel     Middle East  2012            3.4                4.3   

   Adult_mortality  Alcohol_consumption  Hepatitis_B  Measles   BMI  ...  \
0         105.8240                 1.32           97       65  27.8  ...   
1          57.9025                10.35           97       94  26.0  ...   
2         201.0765                 1.57           60       35  21.2  ...   
3         222.1965                 5.68           93       74  25.3  ...   
4          57.9510                 2.89           97       89  27.0  ...   

   Diphtheria  Incidents_HIV  GDP_per_capita  Population_mln  \
0          97           0.08           11006           78.53   
1          97           0.09           25742           46.44   
2          64           0.13            1076         1183.21   
3          93           0.79            4146            0.75   
4          94           0.08           33995            7.91   

   Thinness_ten_nineteen_years  Thinness_five_nine_years  Schooling  \
0                          4.9                       4.8        7.8   
1                          0.6                       0.5        9.7   
2                         27.1                      28.0        5.0   
3                          5.7                       5.5        7.9   
4                          1.2                       1.1       12.8   

   Economy_status_Developed  Economy_status_Developing  Life_expectancy  
0                         0                          1             76.5  
1                         1                          0             82.8  
2                         0                          1             65.4  
3                         0                          1             67.0  
4                         1                          0             81.7  

[5 rows x 21 columns]

## Split

In [5]:
## Remove y from list
feature_cols = list(df.columns)
feature_cols.remove('Life_expectancy')

In [7]:
# Create X, and y.

X = df[feature_cols]  ## independant variable or feature
y = df['Life_expectancy'] ## dependant variable or target

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 22) ## random state 22

In [9]:
(X_train.index==y_train.index).sum() ## Checking X & y trains are the same row count

2291

In [10]:
X_train.shape[0]  ## Every single index between X_train and y_train AND THE ORDER OF THEM

2291

# Feature Engineering

In [14]:
def feature_eng(df):
    df = df.copy()

    '''one hot encoding for Region'''
    df = pd.get_dummies(df, columns = ['Region'], drop_first=True, prefix = 'Region')

    '''Drop Country'''
    df=df.drop(columns=['Country'])

    '''adding the constant for the intercept'''
    df = sm.add_constant(df)

    return df

In [15]:
X_train_fe = feature_eng(X_train)

In [16]:
X_train_fe.head(10)

const  Year  Infant_deaths  Under_five_deaths  Adult_mortality  \
2453    1.0  2003           13.4               15.6         153.4715   
75      1.0  2014            3.1                4.1          74.8905   
1736    1.0  2011           24.5               29.3         141.3175   
1925    1.0  2012           35.6               44.4         137.2305   
120     1.0  2014           46.0               68.3         309.5495   
2783    1.0  2007           52.0               77.9         258.5945   
1186    1.0  2007           43.0               61.9         312.3995   
2186    1.0  2013            8.2                9.6          88.8035   
2809    1.0  2011           69.9               96.7         262.7215   
1332    1.0  2000           43.9               52.3         152.9405   

      Alcohol_consumption  Hepatitis_B  Measles   BMI  Polio  ...  \
2453                 1.79           85       96  31.4     98  ...   
75                  12.07           83       77  25.0     98  ...   
1736                 1.82           63       65  26.0     75  ...   
1925                 0.00           94       41  21.3     94  ...   
120                  4.22           86       33  22.4     78  ...   
2783                 0.64           84       64  20.9     83  ...   
1186                 1.75           91       88  20.4     95  ...   
2186                 3.02           94       92  27.1     95  ...   
2809                 3.73           77       64  23.9     77  ...   
1332                 0.45           43       92  24.6     95  ...   

      Economy_status_Developed  Economy_status_Developing  Region_Asia  \
2453                         0                          1            0   
75                           1                          0            0   
1736                         0                          1            0   
1925                         0                          1            1   
120                          0                          1            0   
2783                         0                          1            0   
1186                         0                          1            0   
2186                         0                          1            0   
2809                         0                          1            0   
1332                         0                          1            0   

      Region_Central America and Caribbean  Region_European Union  \
2453                                     0                      0   
75                                       0                      1   
1736                                     0                      0   
1925                                     0                      0   
120                                      0                      0   
2783                                     0                      0   
1186                                     0                      0   
2186                                     1                      0   
2809                                     0                      0   
1332                                     0                      0   

      Region_Middle East  Region_North America  Region_Oceania  \
2453                   0                     0               1   
75                     0                     0               0   
1736                   0                     0               1   
1925                   0                     0               0   
120                    0                     0               0   
2783                   0                     0               0   
1186                   0                     0               0   
2186                   0                     0               0   
2809                   0                     0               0   
1332                   0                     0               0   

      Region_Rest of Europe  Region_South America  
2453                      0                     0  
75                        0                    

In [76]:
X_test.head()

Country          Region  Year  Infant_deaths  Under_five_deaths  \
1296  Slovak Republic  European Union  2009            5.9                7.2   
2160          Ecuador   South America  2005           19.9               23.5   
2218           Uganda          Africa  2008           54.5               86.8   
1534             Fiji         Oceania  2012           20.0               23.5   
1820            Malta  European Union  2012            5.8                6.6   

      Adult_mortality  Alcohol_consumption  Hepatitis_B  Measles   BMI  Polio  \
1296         129.6060                11.13           99       97  26.0     99   
2160         146.2040                 3.90           92       95  25.9     89   
2218         384.1585                11.37           71       64  21.8     69   
1534         191.3835                 2.43           99       94  27.3     99   
1820          54.2180                 7.18           93       91  27.1     99   

      Diphtheria  Incidents_HIV  GDP_per_capita  Population_mln  \
1296          99           0.01           13679            5.39   
2160          92           0.18            4893           13.83   
2218          71           3.44             725           30.43   
1534          99           0.11            4683            0.87   
1820          99           0.08           21222            0.42   

      Thinness_ten_nineteen_years  Thinness_five_nine_years  Schooling  \
1296                          1.3                       1.4       11.4   
2160                          1.4                       1.3        7.3   
2218                          6.2                       6.2        5.2   
1534                          4.0                       3.7       10.1   
1820                          0.8                       0.7       10.8   

      Economy_status_Developed  Economy_status_Developing  
1296                         1                          0  
2160                         0                          1  
2218                         0                          1  
1534                         0                          1  
1820                         1                          0

In [100]:
y_test.head()

1296    74.9
2160    74.1
2218    55.0
1534    66.9
1820    80.7
Name: Life_expectancy, dtype: float64

# Try

### First ranking all independants/features by their corrlation to Life Expectancy

In [60]:
X_train_fe.columns ## Full Column list

Index(['const', 'Year', 'Infant_deaths', 'Under_five_deaths',
       'Adult_mortality', 'Alcohol_consumption', 'Hepatitis_B', 'Measles',
       'BMI', 'Polio', 'Diphtheria', 'Incidents_HIV', 'GDP_per_capita',
       'Population_mln', 'Thinness_ten_nineteen_years',
       'Thinness_five_nine_years', 'Schooling', 'Economy_status_Developed',
       'Economy_status_Developing', 'Region_Asia',
       'Region_Central America and Caribbean', 'Region_European Union',
       'Region_Middle East', 'Region_North America', 'Region_Oceania',
       'Region_Rest of Europe', 'Region_South America'],
      dtype='object')

In [62]:
## Creating the rank

rankedcorr = pd.get_dummies(df, columns = ['Region'], drop_first = True, prefix = 'Region').copy()  ## Using DF due to still having y value, so needed to 1 hot encode region for rankedcorr
rankedcorr=rankedcorr[rankedcorr.columns[1:]].corr()['Life_expectancy'].reset_index()
rankedcorr['Life_expectancy']=rankedcorr['Life_expectancy'].abs()  ## Absolute values means the returned ranked is based on strength of corrolation regardless of positive or negative correlation
rankedcorr=rankedcorr.sort_values(['Life_expectancy'],ascending=False)
rankedcorr

index  Life_expectancy
18                       Life_expectancy         1.000000
3                        Adult_mortality         0.945360
2                      Under_five_deaths         0.920419
1                          Infant_deaths         0.920032
15                             Schooling         0.732484
8                                  Polio         0.641217
9                             Diphtheria         0.627541
7                                    BMI         0.598423
11                        GDP_per_capita         0.583090
10                         Incidents_HIV         0.553027
17             Economy_status_Developing         0.523791
16              Economy_status_Developed         0.523791
6                                Measles         0.490019
13           Thinness_ten_nineteen_years         0.467824
14              Thinness_five_nine_years         0.458166
5                            Hepatitis_B         0.417804
4                    Alcohol_consumption         0.399159
21                 Region_European Union         0.397038
25                 Region_Rest of Europe         0.182325
0                                   Year         0.174359
22                    Region_Middle East         0.158572
20  Region_Central America and Caribbean         0.131239
23                  Region_North America         0.124750
26                  Region_South America         0.111872
19                           Region_Asia         0.026836
12                        Population_mln         0.026298
24                        Region_Oceania         0.018001

### Using statsmodel, running Linear Regression

In [32]:
feature_cols = ['const', 'Adult_mortality','Under_five_deaths','Infant_deaths','Incidents_HIV','Economy_status_Developing']


lin_reg = sm.OLS(y_train, X_train_fe[feature_cols])
results = lin_reg.fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Life_expectancy   R-squared:                       0.976
Model:                            OLS   Adj. R-squared:                  0.976
Method:                 Least Squares   F-statistic:                 1.896e+04
Date:                Mon, 10 Jul 2023   Prob (F-statistic):               0.00
Time:                        11:47:53   Log-Likelihood:                -4096.0
No. Observations:                2291   AIC:                             8204.
Df Residuals:                    2285   BIC:                             8238.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
=============================================================================================
                                coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------
const                        83.6456      0.087    963.277      0.000      83.475      83.816
Adult_mortality              -0.0477      0.001    -72.858      0.000      -0.049      -0.046
Under_five_deaths            -0.0488      0.004    -11.046      0.000      -0.057      -0.040
Infant_deaths                -0.0686      0.007     -9.513      0.000      -0.083      -0.054
Incidents_HIV                 0.0953      0.020      4.793      0.000       0.056       0.134
Economy_status_Developing    -1.9567      0.090    -21.761      0.000      -2.133      -1.780
==============================================================================
Omnibus:                        0.318   Durbin-Watson:                   2.025
Prob(Omnibus):                  0.853   Jarque-Bera (JB):                0.306
Skew:                          -0.028   Prob(JB):                        0.858
Kurtosis:                       3.003   Cond. No.                         808.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Calculate Train RMSE

In [102]:
## Generate the predictions

y_pred = results.predict(X_train_fe[feature_cols]) ## Predict on the trianing set with the selected features

rmse = statsmodels.tools.eval_measures.rmse(y_train, y_pred) ## Evaluate the RMSE

## Print the RMSE
print(rmse)

1.4461638332250788


# Test

In [35]:
X_test_fe = feature_eng(X_test)
X_test_fe = X_test_fe[feature_cols]


### Calculate Test RMSE

In [103]:
y_test_pred = results.predict(X_test_fe)
rmse_Test = statsmodels.tools.eval_measures.rmse(y_test, y_test_pred)
print(rmse_Test)

1.391260545505006


# Limited Model
For clarity, everything for the model will be seperated from the main model with a 'L_'^

## Try

In [45]:
L_feature_cols = ['const', 'Adult_mortality','Under_five_deaths','Infant_deaths','Economy_status_Developing']


L_lin_reg = sm.OLS(y_train, X_train_fe[L_feature_cols])
L_results = L_lin_reg.fit()
L_results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:        Life_expectancy   R-squared:                       0.976
Model:                            OLS   Adj. R-squared:                  0.976
Method:                 Least Squares   F-statistic:                 2.347e+04
Date:                Mon, 10 Jul 2023   Prob (F-statistic):               0.00
Time:                        12:11:37   Log-Likelihood:                -4107.4
No. Observations:                2291   AIC:                             8225.
Df Residuals:                    2286   BIC:                             8254.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
=============================================================================================
                                coef    std err          t      P>|t|      [0.025      0.975]
---------------------------------------------------------------------------------------------
const                        83.4542      0.077   1077.150      0.000      83.302      83.606
Adult_mortality              -0.0454      0.000   -101.345      0.000      -0.046      -0.044
Under_five_deaths            -0.0488      0.004    -10.997      0.000      -0.057      -0.040
Infant_deaths                -0.0731      0.007    -10.169      0.000      -0.087      -0.059
Economy_status_Developing    -1.9928      0.090    -22.134      0.000      -2.169      -1.816
==============================================================================
Omnibus:                        0.564   Durbin-Watson:                   2.017
Prob(Omnibus):                  0.754   Jarque-Bera (JB):                0.627
Skew:                           0.020   Prob(JB):                        0.731
Kurtosis:                       2.930   Cond. No.                         793.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

### Limited model Train RMSE

In [106]:
L_y_pred = L_results.predict(X_train_fe[L_feature_cols])

L_rmse = statsmodels.tools.eval_measures.rmse(y_train, L_y_pred)

print(L_rmse)

1.4534145336653435


### Limited Model Test

In [47]:
L_X_test_fe = X_test_fe[L_feature_cols]

In [52]:
L_X_test_fe.head()

const  Adult_mortality  Under_five_deaths  Infant_deaths  \
1296    1.0         129.6060                7.2            5.9   
2160    1.0         146.2040               23.5           19.9   
2218    1.0         384.1585               86.8           54.5   
1534    1.0         191.3835               23.5           20.0   
1820    1.0          54.2180                6.6            5.8   

      Economy_status_Developing  
1296                          0  
2160                          1  
2218                          1  
1534                          1  
1820                          0

In [107]:
L_y_test_pred = L_results.predict(L_X_test_fe)
L_rmse_Test = statsmodels.tools.eval_measures.rmse(y_test, L_y_test_pred)
print(L_rmse_Test)

1.4013873399183152


# Final Prep

### All the RMSE's together

In [105]:
print('Train RMSE=',rmse)
print('Test RMSE=',rmse_Test)
print('And for the limited model:')
print('Train RMSE=',L_rmse)
print('Test RMSE=',L_rmse_Test)

Train RMSE= 1.4461638332250788
Test RMSE= 1.391260545505006
And for the limited model:
Train RMSE= 1.4534145336653435
Test RMSE= 1.4013873399183152


### Saving the models

In [85]:
import joblib

# Save the trained model
joblib.dump(results, 'linear_regression_model.pkl')
joblib.dump(L_results, 'limited_model.pkl')

['limited_model.pkl']

# TTS & Modelling Finished
Head to the next workbook for the final function & the input
Project 3.3  - Final.ipynb